File: 240619/최소 차 분할/minimum-diff-partition.py
n = int(input())
l = [0] + list(map(int, input().split()))
total = sum(l)

dp = [ [False] * (total + 1) for _ in range(n+1)]
dp[0][0] = True

ans = total 

for i in range(1, n+1):
    for j in range(total+1):
        if j >= l[i] and dp[i-1][j-l[i]]:
            dp[i][j] = True
        if dp[i-1][j]:
            dp[i][j] = True 

for i in range(1, total):
    if dp[n][i]:
        ans = min(ans, abs(i - (total - i)))
print(ans)

File: 241025/최소 차 분할/minimum-diff-partition.java
import java.util.*;
public class Main {
    static int MAX_N = 101;
    static int MAX_M = 100001;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int[] arr = new int[MAX_N];
        for(int i = 1; i <= N; i++){
            arr[i] = sc.nextInt();
        }

        int total = 0;
        for(int i = 1; i <= N; i++){
            total += arr[i];
        }

        boolean[][] dp = new boolean[MAX_N][MAX_M];
        dp[0][0] = true;
        for(int i = 1; i<= N; i++){
            int v = arr[i];
            for(int j = 0; j < MAX_M; j++){
                if(v <= j && dp[i-1][j-v]) dp[i][j] = true;
                if(dp[i-1][j]) dp[i][j] = true;
            }
        }

        int ans = total;
        for(int j = 0; j <= total; j++){
            if(dp[N][j]) ans = Math.min(ans, Math.abs(j - (total - j)));
        }
        System.out.println(ans);
    }
}